### PR TITLE
[Core] Disabling object store from appearing in worker coredumps

### LINKED ISSUE #29576: [Core] SIGABRT causes Ray to hang
### What happened + What you expected to happen

The below script causes Ray to hang. I would expect to see a `RayActorError` instead.

The output when running the script

```
(base) ray@ip-172-31-80-95:~/default$ python test.py 
2022-10-21 17:56:17,484 INFO worker.py:1230 -- Using address localhost:9031 set in the environment variable RAY_ADDRESS
2022-10-21 17:56:17,742 INFO worker.py:1342 -- Connecting to existing Ray cluster at address: 172.31.80.95:9031...
2022-10-21 17:56:17,749 INFO worker.py:1519 -- Connected to Ray cluster. View the dashboard at https://console.anyscale-staging.com/api/v2/sessions/ses_tU65uDZ9uHVfEDuQzRnsX3EY/services?redirect_to=dashboard 
2022-10-21 17:56:17,751 INFO packaging.py:354 -- Pushing file package 'gcs://_ray_pkg_eff56cd0e859ca3470d515d0dc41effb.zip' (0.00MiB) to Ray cluster...
2022-10-21 17:56:17,752 INFO packaging.py:367 -- Successfully pushed file package 'gcs://_ray_pkg_eff56cd0e859ca3470d515d0dc41effb.zip'.
(Actor pid=11471) *** SIGABRT received at time=1666400179 on cpu 14 ***
(Actor pid=11471) PC: @     0x7f70d75e000b  (unknown)  raise
(Actor pid=11471)     @     0x7f70d77a3420  (unknown)  (unknown)
(Actor pid=11471)     @ ... and at least 3 more frames
(Actor pid=11471) [2022-10-21 17:56:19,050 E 11471 11471] logging.cc:361: *** SIGABRT received at time=1666400179 on cpu 14 ***
(Actor pid=11471) [2022-10-21 17:56:19,050 E 11471 11471] logging.cc:361: PC: @     0x7f70d75e000b  (unknown)  raise
(Actor pid=11471) [2022-10-21 17:56:19,050 E 11471 11471] logging.cc:361:     @     0x7f70d77a3420  (unknown)  (unknown)
(Actor pid=11471) [2022-10-21 17:56:19,050 E 11471 11471] logging.cc:361:     @ ... and at least 3 more frames
(Actor pid=11471) Fatal Python error: Aborted
(Actor pid=11471) 
(Actor pid=11471) Stack (most recent call first):
(Actor pid=11471)   File "test.py", line 32 in abort
(Actor pid=11471)   File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/util/tracing/tracing_helper.py", line 466 in _resume_span
(Actor pid=11471)   File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/_private/function_manager.py", line 674 in actor_method_executor
(Actor pid=11471)   File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/_private/worker.py", line 763 in main_loop
(Actor pid=11471)   File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/_private/workers/default_worker.py", line 231 in <module>
(raylet) [2022-10-21 17:57:27,057 E 761 802] (raylet) file_system_monitor.cc:105: /tmp/ray/session_2022-10-21_17-03-26_460534_151 is over 95% full, available space: 15603879936; capacity: 312201752576. Object creation will fail if spilling is required.
```

### Versions / Dependencies

 master

### Reproduction script

```python
import os
import ray

@ray.remote
class Actor:
    def abort(self):
        os.abort()
        

a = Actor.remote()
ray.get(a.abort.remote())
```

### Issue Severity

High: It blocks me from completing my task.

### PR DESCRIPTION
**Pull request by cadedaniel**
## Description
If a Ray worker process crashes and generates a coredump, Linux will by default dump all pages mapped by the process. This includes the plasma pages, which on large instances can be quite large. This PR uses `madvise` to disable dumping of the plasma pages in worker processes. The impact is that coredumps generated by Ray worker processes are now ~300MB instead of ~= object store size.

See this [[public] Coredumps appearing to cause hangs in Ray](https://docs.google.com/document/d/1_FRz7lm0F4JrzaIjOCBCxE_P7pkjoXYHzFhRDPE8JjM/edit#) for more information.

## Testing
It is difficult to test this in CI because 1) at the C++ level there isn't a Linux API to verify `madvise` status of pages and 2) coredumps aren't enabled (`/proc/sys/kernel/core_pattern` is set to an invalid value and `ulimit -c` is 0) so a Python-level test won't work. I can go and enable coredumps in CI but that feels like a big change, want to check before going down that path.

In terms of manual testing, this is disabled by macro for non Linux builds. On Linux, the coredump size goes down significantly:

```
$ ls -alh /tmp/core.ray::Actor.abor.88940 # without madvise
-rw------- 1 ray users 9.6G Nov 13 10:52 /tmp/core.ray::Actor.abor.88940
$ ls -alh /tmp/core.ray::Actor.abor.97217 # with madvise
-rw------- 1 ray users 239M Nov 13 11:09 /tmp/core.ray::Actor.abor.97217
$ gdb -c /tmp/core.ray::Actor.abor.97217
(gdb) info proc mappings
Mapped address spaces:

          Start Addr           End Addr       Size     Offset objfile
      0x55a3af4a7000     0x55a3af506000    0x5f000        0x0 /home/ray/anaconda3/bin/python3.8
      0x55a3af506000     0x55a3af6fe000   0x1f8000    0x5f000 /home/ray/anaconda3/bin/python3.8
      0x55a3af6fe000     0x55a3af7e5000    0xe7000   0x257000 /home/ray/anaconda3/bin/python3.8
      0x55a3af7e6000     0x55a3af7eb000     0x5000   0x33e000 /home/ray/anaconda3/bin/python3.8
      0x55a3af7eb000     0x55a3af823000    0x38000   0x343000 /home/ray/anaconda3/bin/python3.8
      0x7fc9792c0000     0x7fcd64000000 0x3ead40000        0x0 /dev/shm/plasmax6oDcM (deleted)
(...)
```

Closes https://github.com/ray-project/ray/issues/29576

### CODE REVIEW

File: python/ray/_private/workers/default_worker.py [hunk 1]
```diff
@@ -229,6 +229,8 @@
     )
     configure_log_file(out_file, err_file)
 
+    utils.maybe_disable_object_store_coredumps()
```
Comment (ericl) [inline python/ray/_private/workers/default_worker.py:232]: Could we add a test that the filter is set correctly for a task?

Comment (cadedaniel) [reply python/ray/_private/workers/default_worker.py:232]: For sure! I wanted to make sure this was the right place for this kind of change before fleshing it out. Any high-level feedback on this approach?

Comment (ericl) [reply python/ray/_private/workers/default_worker.py:232]: Lgtm if it works.

**Review by ericl — APPROVED** [review]

File: src/ray/common/ray_config_def.h [hunk 2]
```diff
@@ -710,3 +710,6 @@ RAY_CONFIG(int64_t, health_check_initial_delay_ms, 5000)
 RAY_CONFIG(int64_t, health_check_period_ms, 3000)
 RAY_CONFIG(int64_t, health_check_timeout_ms, 10000)
 RAY_CONFIG(int64_t, health_check_failure_threshold, 5)
+
+/// Use madvise to prevent worker coredump from including the mapped plasma pages.
+RAY_CONFIG(bool, plasma_client_madvise_dontdump, true)
```
Comment (scv119) [inline src/ray/common/ray_config_def.h:715]: nit: core_dump_exclude_plasma_store?

Comment (scv119) [reply src/ray/common/ray_config_def.h:715]: should we make it optional for raylet as well?

**Comment by scv119** [comment]
TIL madvise

File: src/ray/common/ray_config_def.h (hunk 2 above)
Comment (cadedaniel) [reply src/ray/common/ray_config_def.h:715]: Do we need to turn it off for the raylet as well? My impression was that if the raylet crashes, then the object store memory contents could be more significant. So best to leave on until it impacts a use case.

Comment (cadedaniel) [reply src/ray/common/ray_config_def.h:715]: I disabled it for raylet as well but removed the code

Comment (cadedaniel) [reply src/ray/common/ray_config_def.h]: discussed offline, I will disable the same for raylet in a separate PR

**Review by scv119 — APPROVED** [review]

**Comment by cadedaniel** [comment]
Test failures are unrelated. I checked:
* Apple Wheels and Jars
* Apple C++ Unit Tests
* Minimal Install 3.7
* Minimal Install 3.8
* Java
* Dashbaord Tests
* Post-wheel tests

**Pull request merged**